File: DotNetOnion/Cells/CellNetInfo.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using DotNetOnion.Helpers;

namespace DotNetOnion.Cells
{
    public class CellNetInfo : Cell
    {
        public uint Time { get; set; }
        public ORAddress OtherAddress { get; set; }
        public List<ORAddress> MyAddresses { get; set; } = new List<ORAddress>();

        public override byte Command => 0x08;

        public override void Deserialize(BinaryReader reader)
        {
            Time = reader.ReadUInt32BigEndian();
            OtherAddress = ReadORAddress(reader);
            int myAddressCount = reader.ReadByte();
            for (int i = 0; i < myAddressCount; i++)
            {
                MyAddresses.Add(ReadORAddress(reader));
            }
        }

        private ORAddress ReadORAddress (BinaryReader reader)
        {
            return new ORAddress
            {
                Type = reader.ReadByte(),
                Value = reader.ReadBytes(reader.ReadByte())
            };
        }

        private void WriteORAddress(ORAddress address, BinaryWriter writer)
        {
            writer.Write(address.Type);
            writer.Write((byte)address.Value.Length);
            writer.Write(address.Value);
        }

        public override void Serialize(BinaryWriter writer)
        {
            writer.WriteUInt32BigEndian(Time);
            WriteORAddress(OtherAddress, writer);
            writer.Write((byte)MyAddresses.Count);
            foreach (var address in MyAddresses)
                WriteORAddress(address, writer);
        }

        public class ORAddress
        {
            public byte Type { get; set; }
            public byte[] Value { get; set; }
        }

    }
}

File: NOnion/Cells/CellNetInfo.fs
namespace NOnion.Cells

open System.IO

open FSharpx.Collections

open NOnion.Utility.BinaryIO

type RouterAddress =
    {
        Type: byte
        Value: array<byte>
    }

type CellNetInfo =
    {
        Time: uint32
        MyAddresses: seq<RouterAddress>
        OtherAddress: RouterAddress
    }

    static member Deserialize(reader: BinaryReader) =

        let readAddress() : RouterAddress =
            {
                RouterAddress.Type = reader.ReadByte()
                Value = reader.ReadByte() |> int |> reader.ReadBytes
            }

        let rec readAddresses addresses remainingCount =
            if remainingCount = 0uy then
                addresses
            else
                readAddresses
                    (addresses @ [ readAddress() ])
                    (remainingCount - 1uy)

        let time = ReadBigEndianUInt32 reader
        let otherAddress = readAddress()
        let myAddressesCount = reader.ReadByte()
        let myAddresses = readAddresses List.Empty myAddressesCount

        {
            Time = time
            MyAddresses = myAddresses
            OtherAddress = otherAddress
        }
        :> ICell

    interface ICell with

        member __.Command = 8uy

        member self.Serialize writer =

            let writeAddress(addr: RouterAddress) =
                writer.Write addr.Type
                addr.Value.Length |> byte |> writer.Write
                writer.Write addr.Value

            let rec writeAddresses(addresses: seq<RouterAddress>) =
                match Seq.tryHeadTail addresses with
                | None -> ()
                | Some(addr, nextAddresses) ->
                    writeAddress addr
                    writeAddresses nextAddresses

            WriteUInt32BigEndian writer self.Time
            writeAddress self.OtherAddress
            self.MyAddresses |> Seq.length |> byte |> writer.Write
            writeAddresses self.MyAddresses
